tests/components/iskra/test_config_flow.py

```diff
@@ -0,0 +1,241 @@
+    assert result["errors"] == {"base": "invalid_auth"}
+    assert result["step_id"] == "authentication"
+
+    # Test successful authentication
+    mock_pyiskra_rest.side_effect = None
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        user_input={CONF_USERNAME: "", CONF_PASSWORD: ""},
+    )
+
+    # Test successful Rest API configuration
+    assert result["type"] is FlowResultType.CREATE_ENTRY
+    assert result["title"] == f"{SG_MODEL} {SERIAL}"
+    assert result["data"][CONF_HOST] == HOST
+
+
+async def test_user_modbus(hass: HomeAssistant, mock_pyiskra_modbus) -> None:
+    """Test the user flow with Modbus TCP protocol."""
+
+    result = await hass.config_entries.flow.async_init(
+        "iskra",
+        context={"source": SOURCE_USER},
+    )
+
+    assert result["type"] is FlowResultType.FORM
+    assert result["step_id"] == "user"
+
+    result = await hass.config_entries.flow.async_init(
+        "iskra",
+        context={"source": SOURCE_USER},
+        data={CONF_HOST: HOST, CONF_PROTOCOL: "modbus_tcp"},
+    )
+
+    # Test if propmpted to enter port and address
+    assert result["type"] is FlowResultType.FORM
+    assert result["step_id"] == "modbus_tcp"
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        user_input={
+            CONF_PORT: MODBUS_PORT,
+            CONF_ADDRESS: MODBUS_ADDRESS,
+        },
+    )
+
+    # Test successful Modbus TCP configuration
+    assert result["type"] is FlowResultType.CREATE_ENTRY
+    assert result["title"] == f"{PQ_MODEL} {SERIAL}"
+    assert result["data"][CONF_HOST] == HOST
+
+
+async def test_abort_if_already_setup(hass: HomeAssistant, mock_pyiskra_rest) -> None:
+    """Test we abort if Iskra is already setup."""
+
+    MockConfigEntry(domain="iskra", unique_id=SERIAL).add_to_hass(hass)
+    result = await hass.config_entries.flow.async_init(
+        "iskra",
+        context={"source": SOURCE_USER},
+        data={CONF_HOST: HOST, CONF_PROTOCOL: "rest_api"},
+    )
+
+    assert result["type"] is FlowResultType.ABORT
+    assert result["reason"] == "already_configured"
+
+
+@pytest.mark.parametrize(
+    ("s_effect", "reason"),
+    [
+        (DeviceConnectionError, "cannot_connect"),
+        (DeviceTimeoutError, "cannot_connect"),
+        (InvalidResponseCode, "cannot_connect"),
+        (Exception, "unknown"),
+    ],
+)
+async def test_device_abort_rest(
+    hass: HomeAssistant, mock_pyiskra_rest, s_effect, reason
+) -> None:
+    """Test device abort with Rest API protocol."""
+    mock_pyiskra_rest.side_effect = s_effect
+
+    result = await hass.config_entries.flow.async_init(
+        "iskra",
+        context={"source": SOURCE_USER},
+        data={CONF_HOST: HOST, CONF_PROTOCOL: "rest_api"},
+    )
+
+    # Test if aborted
+    assert result["type"] is FlowResultType.FORM
+    assert result["errors"] == {"base": reason}
+    assert result["step_id"] == "user"
```

Comment:
Tests should end in either create_entry or abort. So let's repatch this test and finalize it so we also test this is testing that the flow is able to end

Reply:
I parametrized this test so it's same for restapi and modbus. I Also changed test name as it was misleading, i wanted to test if test returns error, which doesn't cause abort. Should i add code that after error is tested to proceed with normal error free creation of the device?

Reply:
Yes! This way we test that the flow is able to recover and eliminate the possibility of introducing a stateful bug where that can't happen anymore